libsql-server/src/replication/primary/replication_logger_wal.rs

```diff
@@ -0,0 +1,230 @@
+        self.inner.find_frame(page_no)
+    }
+
+    fn read_frame(&mut self, frame_no: u32, buffer: &mut [u8]) -> Result<()> {
+        self.inner.read_frame(frame_no, buffer)
+    }
+
+    fn db_size(&self) -> u32 {
+        self.inner.db_size()
+    }
+
+    fn begin_write_txn(&mut self) -> Result<()> {
+        self.inner.begin_write_txn()
+    }
+
+    fn end_write_txn(&mut self) -> Result<()> {
+        self.inner.end_write_txn()
+    }
+
+    fn undo(
+        &mut self,
+        cb: Option<unsafe extern "C" fn(*mut c_void, u32) -> i32>,
+        cb_ctx: *mut c_void,
+    ) -> Result<()> {
+        self.rollback();
+        self.inner.undo(cb, cb_ctx)
+    }
+
+    fn savepoint(&mut self, rollback_data: &mut [u32]) {
+        self.inner.savepoint(rollback_data)
+    }
+
+    fn savepoin_undo(&mut self, rollback_data: &mut [u32]) -> Result<()> {
+        self.inner.savepoin_undo(rollback_data)
+    }
+
+    fn insert_frames(
+        &mut self,
+        page_size: c_int,
+        page_headers: *mut PgHdr,
+        size_after: u32,
+        is_commit: bool,
+        sync_flags: c_int,
+    ) -> Result<()> {
+        assert_eq!(page_size, 4096);
+        let mut frame_count = 0;
+        for (page_no, data) in PageHdrIter::new(page_headers, page_size as _) {
+            self.write_frame(page_no, data);
+            frame_count += 1;
+        }
+        if let Err(e) = self.flush(size_after) {
+            tracing::error!("error writing to replication log: {e}");
+            // returning IO_ERR ensure that xUndo will be called by sqlite.
+            return Err(rusqlite::ffi::Error::new(SQLITE_IOERR));
+        }
+
+        self.inner.insert_frames(page_size, page_headers, size_after, is_commit, sync_flags)?;
+
+        if is_commit {
+            if let Err(e) = self.commit() {
+                // If we reach this point, it means that we have committed a transaction to sqlite wal,
+                // but failed to commit it to the shadow WAL, which leaves us in an inconsistent state.
+                tracing::error!(
+                    "fatal error: log failed to commit: inconsistent replication log: {e}"
+                );
+                std::process::abort();
+            }
+
+            if let Err(e) = self.logger.log_file.write().maybe_compact(
+                self.logger.compactor().clone(),
+                size_after,
+                self.logger.db_path(),
+            ) {
+                tracing::error!("fatal error: {e}, exiting");
+                std::process::abort()
+            }
+        }
+
+        Ok(())
+    }
+
+    fn checkpoint<B: sqld_libsql_bindings::wal::BusyHandler>(
+        &mut self,
+        db: *mut sqlite3,
+        mode: sqld_libsql_bindings::wal::CheckpointMode,
+        busy_handler: Option<&mut B>,
+        sync_flags: u32,
+        // temporary scratch buffer
+        buf: &mut [u8],
+    ) -> Result<(u32, u32)> {
+        self.inner.checkpoint(db, mode, busy_handler, sync_flags, buf)
```

Comment:
so the idea here is that bottomless will implement its own logger, and it will be composed here? How is it going to be composed?